ert3/stats/_stats.py

```diff
@@ -2,85 +2,67 @@
             )
         if size is not None and index is not None:
             raise ValueError(
-                "Cannot create gaussian distribution with both size and index"
+                "Cannot create distribution with both size and index"
             )
 
-        self._index = index
         if size is not None:
             self._size = size
+            self._as_array = True
+            self._index = tuple(range(self._size))
         else:
             self._size = len(tuple(index))
-            self._index = tuple(index)
+            self._as_array = False
+            self._index = index
+
+    @property
+    def index(self):
+        return self._index
+
+    def _add_index(self, x):
+        if self._as_array:
+            return x 
+        else:
+            return {idx: float(val) for idx, val in zip(self.index, x)}
+
+
+class Gaussian(Distribution):
+    def __init__(self, mean, std, *, size=None, index=None):
+        super().__init__(size=size, index=index)
+        self._mean = mean
+        self._std = std
 
     def sample(self):
         sample = scipy.stats.norm.rvs(loc=self._mean, scale=self._std, size=self._size)
-        if self._index is None:
-            return sample
-        else:
-            return {idx: float(val) for idx, val in zip(self._index, sample)}
+        return self._add_index(sample)
 
     def ppf(self, x):
         x = np.full(self._size, x)
         result = scipy.stats.norm.ppf(x, loc=self._mean, scale=self._std)
-        if self._index is None:
-            return result
-        else:
-            return {idx: float(val) for idx, val in zip(self._index, result)}
-
-    @property
-    def index(self):
-        return self._index if self._index is not None else tuple(range(self._size))
+        return self._add_index(result)
 
 
-class Uniform:
+class Uniform(Distribution):
     def __init__(self, lower_bound, upper_bound, *, size=None, index=None):
+        super().__init__(size=size, index=index)
         self._lower_bound = lower_bound
         self._upper_bound = upper_bound
         self._scale = upper_bound - lower_bound
 
-        if size is None and index is None:
-            raise ValueError(
-                "Cannot create uniform distribution with neither size nor index"
-            )
-        if size is not None and index is not None:
-            raise ValueError(
-                "Cannot create uniform distribution with both size and index"
-            )
-
-        if size is not None:
-            self._size = size
-        else:
-            self._size = len(tuple(index))
-        self._index = index
-
     def sample(self):
         sample = scipy.stats.uniform.rvs(
             loc=self._lower_bound, scale=self._scale, size=self._size
         )
-        if self._index is None:
-            return sample
-        else:
-            return {idx: float(val) for idx, val in zip(self._index, sample)}
+        return self._add_index(sample)
```

Comment:
Would it be possible to give the specific `rvs` function of the distribution to the base class in `__init__` and unifying the sample functions in the base class? And probably same for the `ppf`?

Reply:
That is a good point! I'll test it :)

I was actually looking for something like this, but I was hoping to generalise over the entire distribution class so that I could create an `scipy.stats.my_distribution(some, arguments)` that you could call `rvs` and `ppf` on. But that was unfortunately not possible in a natural manner 🤷

Reply:
Agreed, that would be even nicer! Little bit surprised that they don't provide an interface where you can create specific instances of a distribution and use it like you describe.

Another alternative would be to have abstract functions `rvs` and  `ppf` that the derived classes are expected to implement. Don't really have any strong opinions on which is best.

Reply:
I kind of implemented both 😬 See below ⬇️ 

We can create instances, but then only for sampling single values... Sampling these values one by one in Python really kills efficiency, so we'll have to revert it if we were heading down that path.